versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/DynamicVersioningSCMPlugin.java

```diff
@@ -0,0 +1,259 @@
+import org.eclipse.jgit.revwalk.RevWalk;
+import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
+
+/**
+ * Maven plugin that uses SCM/VCS to enable dynamic versioning based on your
+ * version control system.
+ *
+ * @author Jimisola Laursen
+ * @since 2.17.0
+ */
+@Mojo(name = "set-version-from-scm-tag", defaultPhase = LifecyclePhase.INITIALIZE)
+public class DynamicVersioningSCMPlugin extends AbstractMojo {
+
+    /**
+     * The Maven Project Object
+     */
+    @Parameter(defaultValue = "${project}", readonly = true, required = true)
+    private MavenProject project;
+
+    /**
+     * The name of the property that will contain the resolved version.
+     */
+    @Parameter(property = "propertyName", defaultValue = "revision")
+    protected String propertyName;
+    /**
+     * Whether the SNAPSHOT qualifier shall be apppended or not.
+     */
+    @Parameter(property = "appendSnapshot", defaultValue = "true")
+    protected boolean appendSnapshot;
+
+    /**
+     * Set the version instead of resolving from SCM tag information. Environment
+     * variable
+     * <i>DV_SCM_PLUGIN_USE_VERSION<i> can also be used and will
+     * override this
+     * parameter.
+     */
+    @Parameter(property = "useVersion")
+    protected String useVersion;
+
+    /**
+     * The default version used when SCM repository has no commit or no semantic
+     * version
+     * tag.
+     */
+    @Parameter(property = "defaultVersion", defaultValue = "0.0.1")
+    protected String defaultVersion;
+
+    // standard semantic versioning with an optional 'v' prefix
+    protected static final Pattern TAG_VERSION_PATTERN = Pattern.compile("refs/tags/(?:v)?((\\d+\\.\\d+\\.\\d+)(.*))");
+
+    protected static final String DV_SCM_USE_VERSION_ENV_VAR_NAME = "DV_SCM_PLUGIN_USE_VERSION";
+
+    public void execute() throws MojoExecutionException {
+        // limit JGits excessive logging
+        Logger.getLogger("org.eclipse.jgit").setLevel(Level.INFO);
+
+        VersionInformation vi;
+
+        Optional<String> mayBeVersion = getHardcodedVersion();
+
+        if (mayBeVersion.isPresent()) {
+            vi = new VersionInformation(mayBeVersion.get());
+        } else {
+            vi = getVersionFromSCM();
+        }
+
+        project.getProperties().setProperty(propertyName, vi.toString());
+        getLog().info("Property '" + propertyName + "' set to: "
+                + project.getProperties().getProperty(propertyName));
+    }
+
+    /**
+     * Returns value of environment variable
+     * {@code DV_SCM_PLUGIN_USE_VERSION}, parameter
+     * {@code useVersion}
+     * or, if
+     * none are set, Optional.empty() (in that order).
+     *
+     * @return Optional<String> that can contain version set by environment variable
+     *         of
+     *         parameter {@code useVersion}
+     */
+    protected Optional<String> getHardcodedVersion() {
+        String version = System.getenv(DV_SCM_USE_VERSION_ENV_VAR_NAME);
```

Comment:
It is not usual to use environments for plugin configuration.

Why you need it?

Reply:
It is for bypassing the whole SCM functionality. Used e.g. when testing pipelines.

From a GitLab CI/CD Components perspective we could set an environment variable when testing that will add `-DuseVersion=true` to the Maven options.

However, not using an environment variable adds "work" and does not feel as reliable.
This since using `useVersion` from the command line in a Maven build with many plugins there is a risk that `useVersion` is used in other plugins. Whereas the environment variable is more or less guaranteed to be unique.

And option would be to change the parameters so that they are more unique, e.g. like Spring Java Format does with `spring-javaformat.apply.skip`.

The parameter `useVersion` shall not be set in the POM configuration as it main purpose is for testing.

I can remove if it is a deal-breaker but  I think we should think about the naming of the parameters then.

Reply:
Removed.